pyannote/audio/models/blocks/selfsup.py

```diff
@@ -0,0 +1,181 @@
+        super().__init__()
+        if torchaudio_ssl:
+            if checkpoint:
+                raise ValueError("Error : Cannot specify both a checkpoint and a torchaudio model.")
+            
+            print("\nThe Self-Supervised Model "+str(torchaudio_ssl)+" is loaded from torchaudio.\n")
+            name,config,ordered_dict = self.dict_torchaudio(torchaudio_ssl,torchaudio_cache)
+        else:
+            print("A checkpoint from a Self-Supervised Model is used for training.")
+            if torch.cuda.is_available():  
+                ckpt = torch.load(checkpoint)
+            else:
+                ckpt = torch.load(checkpoint,map_location=torch.device('cpu'))
+                #Check if the checkpoint is from an already finetuned Diarization model (containing SSL), or from a SSL pretrained model only
+            if 'pyannote.audio' in ckpt: #1: Check if there is a Segmentation model attached onto or not
+                print("The checkpoint is used for finetuning. \nThe attached SSL model will be used for feature extraction.")
+                name,config,ordered_dict = self.dict_finetune(ckpt)
+
+            else: #Otherwise, load the dictionary of the SSL checkpoint
+                print("The checkpoint is a pretrained SSL model to use for Segmentation.\nBuilding the SSL model.")
+                name,config,ordered_dict = self.dict_pretrained(ckpt)
+                
+        # Layer-wise pooling (same way as SUPERB)
+        if not average_all:      
+            if not average_layers :
+                if layer is None :
+                    print("\nLayer number not specified. Default to layer 1.\n")
+                    
+                    self.layer = 1
+                else :        
+                    
+                    self.layer = layer
+                    print("\nSelected layer is "+ str(layer) +". \n")
+            else:
+                print("Layers "+str(average_layers)+" selected for layer-wise pooling.")
+                
+                self.W = nn.Parameter(torch.randn(len(average_layers))) #Set specific number of learnable weights
+                
+                self.average_layers = average_layers
+                
+                self.layer = max(average_layers)
+        else:
+            print("All layers are selected for layer-wise pooling.")
+            
+            self.W = nn.Parameter(torch.randn(config['encoder_num_layers'])) #Set max number of learnable weights
+            
+            self.average_layers = list(range(config['encoder_num_layers']))
+            
+            self.layer = config['encoder_num_layers']
+            
+        if finetune: #Finetuning not working
+            print("Self-supervised model is unfrozen.")      
+            #config['encoder_ff_interm_dropout'] = 0.3
+            config['encoder_layer_norm_first'] = True
```

Comment:
Can you explain?

Reply:
The issue regarding the funetuning of WavLM seemed similar to a normalization issue that occured during the feature extraction process. If gradient is computed during training, validation will extract feature vectors that are almost identical amongst each frames of the input audio. I assume that it might be the reason why validation does not seem to improve (or change) during training (but I might be completely wrong on this...). Since this problem seemed similar to the one I encountered with WavLM from back a few months (and has been fixed), where features were the also the same amongst the frames, I tried applying a normalization step to see if the behavior of the features extracted would change. Did not seem to be the case... It is one of the many things I tried but forgot to remove when pushing codes ^^